src/output-json-dnp3.c

```diff
@@ -0,0 +1,464 @@
+
+/* Appears not all current distros have jansson that defines this. */
+#ifndef json_boolean
+#define json_boolean(val)      ((val) ? json_true() : json_false())
+#endif
+
+typedef struct LogDNP3FileCtx_ {
+    LogFileCtx *file_ctx;
+    uint32_t    flags;
+    uint8_t     include_object_data;
+} LogDNP3FileCtx;
+
+typedef struct LogDNP3LogThread_ {
+    LogDNP3FileCtx *dnp3log_ctx;
+    uint32_t        count;
+    MemBuffer      *buffer;
+} LogDNP3LogThread;
+
+static int JsonDNP3PrefixIsSize(uint8_t prefix_code)
+{
+    switch (prefix_code) {
+        case 0x04:
+        case 0x05:
+        case 0x06:
+            return 1;
+        default:
+            break;
+    }
+
+    return 0;
+}
+
+static json_t *JsonDNP3LogLinkControl(uint8_t lc)
+{
+    json_t *lcjs = json_object();
+    if (unlikely(lcjs == NULL)) {
+        return NULL;
+    }
+
+    json_object_set_new(lcjs, "dir", json_boolean(DNP3_LINK_DIR(lc)));
+    json_object_set_new(lcjs, "pri", json_boolean(DNP3_LINK_PRI(lc)));
+    json_object_set_new(lcjs, "fcb", json_boolean(DNP3_LINK_FCB(lc)));
+    json_object_set_new(lcjs, "fcv", json_boolean(DNP3_LINK_FCV(lc)));
+    json_object_set_new(lcjs, "function_code", json_integer(DNP3_LINK_FC(lc)));
+
+    return lcjs;
+}
+
+static json_t *JsonDNP3LogIin(uint16_t iin)
+{
+    json_t *iinjs = json_object();
+    if (unlikely(iinjs == NULL)) {
+        return NULL;
+    }
+
+    json_t *indicators = json_array();
+    if (unlikely(indicators == NULL)) {
+        json_decref(iinjs);
+        return NULL;
+    }
+
+    if (iin) {
+        int mapping = 0;
+        do {
+            if (iin & DNP3IndicatorsMap[mapping].value) {
+                json_array_append_new(indicators,
+                    json_string(DNP3IndicatorsMap[mapping].name));
+            }
+            mapping++;
+        } while (DNP3IndicatorsMap[mapping].name != NULL);
+    }
+    json_object_set_new(iinjs, "indicators", indicators);
+
+    return iinjs;
+}
+
+static json_t *JsonDNP3LogApplicationControl(uint8_t ac)
+{
+    json_t *acjs = json_object();
+    if (unlikely(acjs == NULL)) {
+        return NULL;
+    }
+
+    json_object_set_new(acjs, "fir", json_boolean(DNP3_APP_FIR(ac)));
+    json_object_set_new(acjs, "fin", json_boolean(DNP3_APP_FIN(ac)));
+    json_object_set_new(acjs, "con", json_boolean(DNP3_APP_CON(ac)));
+    json_object_set_new(acjs, "uns", json_boolean(DNP3_APP_UNS(ac)));
+    json_object_set_new(acjs, "sequence", json_integer(DNP3_APP_SEQ(ac)));
+
+    return acjs;
+}
+
+/**
+ * \brief Log the items (points) for an object.
+ *
+ * TODO: Autogenerate this function based on object definitions.
+ */
+static json_t *JsonDNP3LogObjectItems(DNP3Object *object)
+{
+    DNP3Point *item;
+    json_t *jsitems;
+
+    if (unlikely((jsitems = json_array()) == NULL)) {
+        return NULL;
+    }
+
+    TAILQ_FOREACH(item, object->points, next) {
+        json_t *js = json_object();
+        if (unlikely(js == NULL)) {
+            break;
```

Comment:
Its still returned even in this case, which means it gets referenced by the parent object, even if empty, so will get free'd with the complete object.

Perhaps that would be more clear if we allocated the json_t object in the caller and passed it in?